pcdsutils/log.py

```diff
@@ -321,3 +338,256 @@ def centralized_logging_enabled() -> bool:
     """Returns True if centralized logging should be enabled."""
     fqdn = get_fully_qualified_domain_name()
     return any(fqdn.endswith(domain) for domain in ALLOWED_LOG_DOMAINS)
+
+
+warnings_logger = logging.getLogger(f'{__name__}.warnings')
+
+
+def log_warning_handler(
+    message: Warning,
+    category: type[Warning],
+    filename: str,
+    lineno: int,
+    file: typing.Optional[typing.TextIO] = None,
+    line: typing.Optional[str] = None,
+    logger: logging.Logger = warnings_logger,
+):
+    """
+    Warning handler that redirects all of the warnings to a logger.
+
+    This can be used as a drop-in replacement for warnings.showwarning to
+    redirect unfiltered warnings into the logging stream.
+
+    Rather than duplicate the warning display text, I've opted to simplify it
+    and put the extra details into the "extra" dictionary argument in the
+    logging library.
+
+    The warnings module displays the warnings as:
+    filename:lineno: category: message\\nline
+    (where, in all cases I've seen, "line" is generated by reading the file)
+
+    The log message generated here will simply be:
+    category: message
+
+    All arguments (except "logger") will be included in the "extra" dictionary.
+    This means they can be used in log filters without parsing the message.
+    The keys used will be "warning_{key}" for each keyword parameter to this
+    function, to avoid collisions.
+
+    Parameters
+    ----------
+    message : Warning
+        This is the Warning object created by a warnings.warn call.
+        When converted using str, this becomes the string message
+        that was passed into warnings.warn. This will be put into the
+        generated log message text and into the extra dict.
+    category : type[Warning]
+        The warning type, e.g. UserWarning, DeprecationWarning, etc.
+        this will be put into the generated log message text and into
+        the extra dict.
+    filename : str
+        The name of the source code file that generated the warning.
+        This will be put into the extra dict.
+    lineno : int
+        The line number in the file that generated the warning. This will
+        be put into the extra dict.
+    file : file-like, optional
+        A file-like object that is normally used in the warnings handler as
+        the destination for warnings, defaulting to sys.stderr. This will
+        be put into the extra dict.
+    line : str, optional
+        The string line in the file that generated the warning. I have never
+        seen this passed into the warning handler. This will be put into
```

Comment:
It's not passed in? Do we want to add it ourselves here?

Reply:
In the testing I did, I never saw a warning come through with `line` set. I'm not sure why this is, maybe it's something you can pass explicitly to `warn`?

I decided not to add it because I didn't want to open an extra file on NFS in a logging handler.

Reply:
I think that's a good reason 👍